src/ipld/cid_vec.rs

```diff
@@ -0,0 +1,178 @@
+// Copyright 2019-2023 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use crate::utils::cid::{CidVariant, BLAKE2B256_SIZE};
+use cid::{
+    multihash::{self, Code::Blake2b256},
+    Cid,
+};
+use fvm_ipld_encoding::DAG_CBOR;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+
+// CidVec takes advantage of the fact that the V1 DAG-CBOR Blake2b-256 variant (which can be stored in 32 bytes vs 96 bytes for a `Cid` type) is +99.99% of all CIDs. CidVec defaults to the `Vec<[u8; BLAKE2B256_SIZE]>` type, only using the more expensive `Vec<Cid>` type when necessary.
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub enum CidVec {
```

Comment:
If I've got a CidVec of length 1000, with just one rare CID, we're going to blow our optimisation.

- how common is this case?
- have we explored e.g
```rust
enum CidNiche {
    Small([u8; 32]),
    Heap(Box<Cid>)
}
```

We could even save a byte on a tagged pointer

Reply:
@lemmih I think I much prefer this:
```rust
use std::mem::size_of;

use cid::Cid;
use static_assertions::const_assert_eq;

#[repr(transparent)]
struct V1DagCborCid([u8; 32]);

const_assert_eq!(size_of::<Cid>(), 96);
const_assert_eq!(size_of::<V1DagCborCid>(), 32);

enum SmallCid {
    Stack(V1DagCborCid),
    Heap(Box<Cid>),
}

const_assert_eq!(size_of::<Box<Cid>>(), 8);
const_assert_eq!(size_of::<SmallCid>(), 40);
```

And we could use `Vec<SmallCid>` or even `SmallVec<SmallCid>` as required...

Reply:
Changed to `Vec<SmallCid>` based on requested changes below.